### PR TITLE
Added a revision number cache to make setJSON safer

### PR DESCRIPTION
**Pull request by redsolver**
## Spec

https://hackmd.io/Sko_QDahR7SP2dQPmAmS-w

### CODE REVIEW

File: src/skydb.ts
Comment (ro-tex) [inline src/skydb.ts:105]: nit: Defaulting to -1 for revision is a little confusing at first. When I thought about it and figured out it was meant to mean "this has never been set" and to make `prev_rev + 1` useful then it made sense. But it would be nice to move it to a `const` and explain the assumptions.

File: src/skydb.ts
Comment (ro-tex) [inline src/skydb.ts:112]: 1. Shouldn't this be `>=` instead of `>`? If we're trying to set revision 5 and that's already cached that means something is wrong.
2. If `cachedRevision` is 0 and we're trying to set 0 again this check won't be triggered.

File: src/skydb.ts
Comment (kwypchlo) [inline src/skydb.ts:394]:
```suggestion
  if (entry === null) {
    return BigInt(0);
  }

  return incrementRevision(entry.revision);
```

File: src/skydb.ts
Comment (kwypchlo) [inline src/skydb.ts:369]: I might be misunderstanding because it goes completely against this change but why are we assuming that we have the registry revision cached if it exists and if not then the registry entry is empty? Previous implementation had explicit `client.registry.getEntry` call to look for the entry just before writing to registry to either pull the version or find out that it doesn't exist. `revisionNumberCache` will only have the value if `getJSON` is called and even if, we're not sure that the cached data is not stale.

I'm going to mark this PR as changes requested even though it's a question but I want to make sure it's blocked until resolved.

**Review by kwypchlo — CHANGES_REQUESTED** [review]

**Comment by ro-tex** [comment]
Is this PR still alive? Let's either close it or push it across the finish line.

File: integration/skydb.test.ts
Comment (m-cat) [reply integration/skydb.test.ts:254]: Yeah, the issue is that if the `expect` fails it tries to serialize both sides to JSON to print them, which runs into a `bigint is not serializable` error.

Comment (kwypchlo) [reply integration/skydb.test.ts:254]: Doesn't seem to be the case when I tested it on jest playground using 24.9.0. Can you double check please?

Screenshot 2021-11-08 at 21 04 41

Comment (m-cat) [reply integration/skydb.test.ts:254]:
```ts
expect(revisionNumber).toEqual(BigInt(1));
```

gives me

```
$ /Users/marcin/Sync/Repos/github.com/SkynetLabs/skynet-js-2/node_modules/.bin/jest -t 'Should update the revision number cache'
node:internal/child_process/serialization:127
    const string = JSONStringify(message) + '\n';
                   ^

TypeError: Do not know how to serialize a BigInt
    at stringify (<anonymous>)
    at writeChannelMessage (node:internal/child_process/serialization:127:20)
    at process.target._send (node:internal/child_process:819:17)
    at process.target.send (node:internal/child_process:719:19)
    at reportSuccess (/Users/marcin/Sync/Repos/github.com/SkynetLabs/skynet-js-2/node_modules/jest-worker/build/workers/processChild.js:67:11)
node:internal/child_process/serialization:127
    const string = JSONStringify(message) + '\n';
                   ^
```

Comment (m-cat) [reply integration/skydb.test.ts:254]: https://github.com/facebook/jest/issues/11617

Comment (kwypchlo) [reply integration/skydb.test.ts:254]: good find, I'll resolve this conversation and maybe we can refactor this once the issue is resolved

**Comment by m-cat** [comment]
We are probably going to change the spec, and I'm still adding tests.

**Comment by m-cat** [comment]
The PR has been updated to protect SkyDB calls with mutexes, as per conversation on Discord. The spec hasn't been updated yet.

Here are the changes since the last review:

- [X] Return errors from skyd instead of status codes
- [X] Fix missing cache updates in `setEntryData` methods
- [X] Add =buildSkylinkJSONObject= method and =SkynetJSON= object per spec
- [X] Add check for multiple simultaneous calls.
- [X] Add tests
  - [X] Integration:
    - [X] setJSON and getJSON called simultaneously (should error)
    - [X] setJSON and getJSON with delays (should error, or getJSON gets new data)
    - [X] two setJSON calls (should error)
    - [X] setJSON and getJSON with two different clients (different local caches)
    - [X] two setJSON calls with two different clients (different local caches)
  - [X] Unit:
    - [X] setJSON and getJSON called simultaneously (should error)
    - [X] two setJSON calls (should error)
    - [X] setJSON and getJSON with two different clients (different local caches)
    - [X] two setJSON calls with two different clients (different local caches)

File: src/mysky/index.ts
Comment (m-cat) [reply src/mysky/index.ts]: Good catch. This case should not occur with the latest changes.

**Review by peterjan — COMMENTED** [review]
looks solid overall

File: integration/skydb.test.ts
Comment (ChrisSchinnerl) [reply integration/skydb.test.ts:261]: Maybe it makes more sense for the `revisionNumberCache` to be a field in the `client.db`. e.g. `client.db.revisionNumberCache` instead. Then it makes sense that calling `db.setJSON` can update the cache because it essentially owns it. While now it appears to be this separate thing.

File: src/skydb.ts
Comment (m-cat) [reply src/skydb.ts:42]: It was not exported in `src/index.ts`, it's only used internally.

File: src/skydb.ts
Comment (m-cat) [reply src/skydb.ts:674]: Fair enough. Not sure how this can come up in production but I added a unit test.

File: integration/skydb.test.ts
Comment (m-cat) [reply integration/skydb.test.ts:304]: > a really good test would assert this path is entered into, can we make that happen easily? Wrapping this in a retry with more concurrent entry access perhaps?

Asserting all paths are entered sounds prone to NDFs to me, unless we have a lot of iterations. Let me know if you think it's important.

> This should really only happen on the 0 delay right? As soon as there's a delay on the getJSON it'll always return jsonNew right?

I think that's only true if setJSON finishes immediately. I don't think that's the case, and there can always arise unpredictable network conditions.

File: integration/skydb.test.ts
Comment (m-cat) [reply integration/skydb.test.ts:348]: Same as above. Kind of tricky case and I don't think I've coded tests like this before, so I don't know what to do.

File: integration/skydb.test.ts
Comment (m-cat) [reply integration/skydb.test.ts:302]: I'm having trouble with this comment, can you give me an example of what you mean? Which cases?

File: integration/skydb.test.ts
Comment (peterjan) [reply integration/skydb.test.ts:373]: I resolved the above conversation. My only point is that usually having a `return` statement in a unit test or integration test hides a possible bug. We could be under the impression that the assertions following this if condition keep passing, but in reality the test might return here 100% of the time.

If you've manually confirmed that is not the case, and we do run the assertions below in a decent amount of cases I guess it's fine. Just wanted to make the remark because I've experienced a situation like this before and since then a `return` in a unit test is a trigger for me.

Comment (peterjan) [reply integration/skydb.test.ts:373]: But you can obviously resolve the comment if you feel it's fine and harmless.

**Review by peterjan — DISMISSED** [review]

**Review by ChrisSchinnerl — DISMISSED** [review]

File: integration/skydb.test.ts
Comment (m-cat) [reply integration/skydb.test.ts:373]: To be fair it is an integration test and not a unit test, and there are two possible real-world conditions that could happen here that we want to explicitly test for. I appreciate that it's not ideal test design, but I don't know how else to write it.

> If you've manually confirmed that is not the case

I have, and will add a comment to that effect.

**Review by peterjan — COMMENTED** [review]
Looks good, would approve if it weren't for the failing test. Seeing as we introduce locking and the test is timing out it warrants investigation as it might be a deadlock, although unlikely.

**Review by peterjan — APPROVED** [review]

**Review by ChrisSchinnerl — APPROVED** [review]

**Pull request merged**
